### PR TITLE
Fix bug 1241251 - Move from peep to pip8.

### PR DESCRIPTION
**Pull request by jezdez**

### CODE REVIEW

File: requirements/compiled.txt [hunk 1]
```diff
@@ -1,17 +1,6 @@
...
-newrelic==2.52.0.40
-
-# sha256: DxedfnXnyDtjQblZXKHzlN5wgUhKnjUq1m1VOhw9qik
-Pillow==2.9.0
+Jinja2==2.7.3 --hash=sha256:2e24ac5d004db5714976a04ac0e80c6df6e47e98c354cb2c0d82f8879d4f8fdb
+MarkupSafe==0.23 --hash=sha256:a4ec1aff59b95a14b45eb2e23761a0179e98319da5a7eb76b56ea8cdc7b871c3
+mysqlclient==1.3.6 --hash=sha256:1f63a8ab111530a0ff9acaeb118074f11e7564b7c124f2e7fc56fcac641aecf2
+lxml==3.4.4 --hash=sha256:b3d362bac471172747cda3513238f115cbd6c5f8b8e6319bf6a97a7892724099
+newrelic==2.52.0.40 --hash=sha256:59b515bb7dbbcdd307f817ac38854e80839ebfa33c5db797f65450793c945c0d
+Pillow==2.9.0 --hash=sha256:0f179d7e75e7c83b6341b9595ca1f394de7081484a9e352ad66d553a1c3daa29
```
Comment (willkg) [inline requirements/compiled.txt]: Hooray!

File: scripts/chief_deploy.py [hunk 2]
```diff
@@ -119,9 +119,9 @@ def setup_dependencies(ctx):
         activate_env = os.path.join(VENV_BIN, 'activate_this.py')
         execfile(activate_env, dict(__file__=activate_env))
 
+        ctx.local('pip install --upgrade pip')
```
Comment (willkg) [inline scripts/chief_deploy.py]: Do we want to lock this down to a specific version of pip?

Comment (jezdez) [inline scripts/chief_deploy.py]: Yeah, I used "<9"

File: requirements/compiled.txt (hunk 1 above)
Comment (jezdez) [inline requirements/compiled.txt]: :boom:

**Comment by willkg** [comment]
I did a quick spin through the code and it looked fine. I'll wait for @jezdez to stop changing things and Travis to greenlight this. Then I'll run it through some vagrant testing.

File: scripts/chief_deploy.py [hunk 3]
```diff
@@ -119,9 +119,13 @@ def setup_dependencies(ctx):
...
         execfile(activate_env, dict(__file__=activate_env))
 
+        ctx.local('pip install --upgrade "pip<9"')
         ctx.local('pip --version')
-        ctx.local('%s scripts/peep.py install -r requirements/default.txt' %
-                  python)
+        ctx.local('pip install '
+                  '--require-hashes '
+                  # until we install compiled dependencies here as well:
+                  '--no-deps '
```
Comment (jezdez) [inline scripts/chief_deploy.py]: to be clear, one of the non-compiled deps requires lxml which is in the compiled.txt but not in default.txt

**Comment by jezdez** [comment]
@willkg r?

**Comment by jezdez** [comment]
(Should we test this on stage before?)

File: provisioning/roles/kuma/tasks/python.yml [hunk 4]
```diff
@@ -2,17 +2,24 @@
...
+    name: pip
+    version: 8.0.0
+    virtualenv: /home/vagrant/env
+
+- name: Install local dependencies in virtualenv
+  pip:
+    requirements: /home/vagrant/src/requirements/local.txt
+    virtualenv: /home/vagrant/env
     virtualenv_site_packages: yes
   with_items:
```
Comment (willkg) [inline provisioning/roles/kuma/tasks/python.yml]: The way this got changed, this `with_items` has no value, so it breaks vagrant provisioning.

**Comment by willkg** [comment]
I fixed the `with_items` thing locally so I could continue testing vagrant provisioning. Now I get this:

```
Installing collected packages: amqp, argparse, anyjson, pytz, Babel, requests, basket-client, BeautifulSoup, billiard, bitly-api, six, html5lib, bleach, kombu, celery, commonware, cssselect, diff-match-patch, Django, python-openid, oauthlib, requests-oauthlib, django-allauth, django-appconf, django-babel, django-cacheback, django-celery, django-celery-transactions, django-constance, django-dbgettext, django-extensions, django-flat-theme, django-honeypot, MarkupSafe, Jinja2, django-jinja, python-memcached, hash-ring, django-memcached-hashring, django-mozilla-product-details, django-picklefield, futures, django-pipeline, django-ratelimit, django-smuggler, django-soapbox, django-statici18n, django-sundial, django-taggit, django-tidings, django-timezones, django-waffle, djangorestframework, urllib3, elasticsearch, python-dateutil, elasticsearch-dsl, feedparser, funcsigs, gunicorn, importlib, jsonpickle, logilab-common, logilab-astng, lxml, pbr, mock, mysqlclient, newrelic, ordereddict, Pillow, polib, puente, pyquery, python-magic, pytidylib, raven, requests-mock, translate-toolkit, URLObject, Werkzeug, whitenoise, coverage, itsdangerous, Flask, greenlet, gevent, msgpack-python, locustio, pep8, pluggy, py, pyflakes, pytest, pytest-cov, pytest-django, virtualenv, tox
  Found existing installation: argparse 1.2.1

:stderr: The directory '/tmp/vagrant-cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/root/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
/home/vagrant/env/local/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:315: SNIMissingWarning: An HTTPS request has been made, but the SNI (Subject Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#snimissingwarning.
  SNIMissingWarning
/home/vagrant/env/local/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:120: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
Detected a distutils installed project ('argparse') which we cannot uninstall. The metadata provided by distutils does not contain a list of files which have been installed, so pip does not know which files to uninstall.


FATAL: all hosts have already failed -- aborting
```

The way I read this, it's a bunch of unfortunate warnings and then ends with an error related to argparse. I figured I'd copy and paste the whole thing in case my reading is wrong.

File: provisioning/roles/kuma/tasks/python.yml (hunk 4 above)
Comment (jezdez) [inline provisioning/roles/kuma/tasks/python.yml]: good catch, this is a rebasing issue

**Comment by jezdez** [comment]
ok, this just means I'll have to stop allowing system packages in the virtualenv

**Comment by jezdez** [comment]
@willkg r?

**Comment by willkg** [comment]
I did a fresh "vagrant up" and it worked great. This looks good to me. r+

I'm going to let @jezdez merge this so that he has full control on when it lands and with what and whether it gets tested on -stage and such because this also includes the "rebuild the virtualenv" changes.

**Comment by jezdez** [comment]
Rebased on master and now for a stage test deploy.

**Comment by jezdez** [comment]
Stage deploy succeeded, took ~4 minutes on the first run with a cold cache, ~3 minutes on a warm HTTP cache. The biggie is lxml compilation, which means we should invest time building whl files for it at some point.

**Comment by jezdez** [comment]
Mergin as it was indicated by @willkg.

**Comment by willkg** [comment]
Hooray! :balloon:

**Pull request merged**